File: Leetcode 3 - longest-substring-without-repeating-characters/3.longest-substring-without-repeating-characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
#include <iostream>
#include <string>
#include <unordered_set>

using namespace std;

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    if (s.empty())
      return 0;

    unordered_set<char> letters;
    int left = 0, right = 0, maxSub = 0;

    while (right < s.size()) {
      if (letters.find(s[right]) == letters.end()) {
        letters.insert(s[right]);
        maxSub = max(maxSub, right - left + 1);
        right++;
      } else {
        letters.erase(s[left]);
        left++;
      }
    }

    return maxSub;
  }
};

int main() {
  Solution sol;

  string s = "abcabcbb";
  cout << sol.lengthOfLongestSubstring(s) << endl; // 3

  s = "bbbbb";
  cout << sol.lengthOfLongestSubstring(s) << endl; // 1

  s = "pwwkew";
  cout << sol.lengthOfLongestSubstring(s) << endl; // 3
}
// @lc code=end

File: Leetcode 3 - longest-substring-without-repeating-characters/3.longest-substring-without-repeating-characters.java
/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start

import java.util.HashSet;
import java.util.Set;

class Solution {

  public int lengthOfLongestSubstring(String s) {
    int left = 0, right = 0, maxCount = 0;

    Set<Character> letters = new HashSet<>();

    while (right < s.length()) {
      if (letters.contains(s.charAt(right))) {
        letters.remove(s.charAt(left));
        left++;
      } else {
        letters.add(s.charAt(right));
        maxCount = Math.max(maxCount, right - left + 1);
        right++;
      }
    }

    return maxCount;
  }

  public static void main(String[] args) {
    Solution solution = new Solution();

    System.out.println(solution.lengthOfLongestSubstring("abcabcbb")); // 3
    System.out.println(solution.lengthOfLongestSubstring("bbbbb")); // 1
    System.out.println(solution.lengthOfLongestSubstring("pwwkew")); // 3
  }
}
// @lc code=end
